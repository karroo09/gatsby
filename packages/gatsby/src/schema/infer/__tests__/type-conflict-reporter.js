const { printConflicts, reportConflict } = require(`../type-conflict-reporter`)

const { getById } = require(`../../db`)
jest.mock(`../../db`, () => {
  const { trackObjects } = require(`../../utils/node-tracking`)

  const nodes = [
    {
      id: 1,
      internal: {
        type: `Foo`,
        description: `Foo Node`,
      },
      bar: 1,
      nested: {
        bar: 1,
      },
    },
    {
      id: 2,
      internal: {
        type: `Foo`,
        description: `Foo Node`,
      },
      bar: true,
      nested: {
        bar: true,
      },
    },
  ]

  nodes.forEach(trackObjects)

  return {
    getById: id => nodes.find(n => n.id === id),
  }
})

const { warn, log } = require(`gatsby-cli/lib/reporter`)
jest.mock(`gatsby-cli/lib/reporter`)

describe(`Type conflict reporter`, () => {
  beforeEach(() => {
    warn.mockReset()
    log.mockReset()
  })

  it(`prints conflicts and their locations`, () => {
    const selector = `Foo.bar`
    const examples = [
      {
        value: 1,
        type: `number`,
        parent: getById(1),
      },
      {
        value: true,
        type: `boolean`,
        parent: getById(2),
      },
    ]

    reportConflict(selector, examples)
    printConflicts()

    expect(warn).toHaveBeenCalledTimes(1)
    expect(warn).toBeCalledWith(
      expect.stringContaining(`There are conflicting field types in your data.`)
    )
    // Could also use Jest's mock serializer, but that also includes return values,
    // which in this case is just noise.
    // expect(log).toMatchSnapshot()
    expect(log.mock.calls[1]).toMatchSnapshot()
    expect(log.mock.calls[2]).toMatchSnapshot()
  })

  it(`prints conflicts and their locations for nested objects`, () => {
    const selector = `Foo.bar.nested`
    const examples = [
      {
        value: 1,
        type: `number`,
        parent: getById(1).nested,
      },
      {
        value: true,
        type: `boolean`,
        parent: getById(2).nested,
      },
    ]

    reportConflict(selector, examples)
    printConflicts()

    expect(warn).toHaveBeenCalledTimes(1)
    expect(log.mock.calls[1]).toMatchSnapshot()
    expect(log.mock.calls[2]).toMatchSnapshot()
  })
})
