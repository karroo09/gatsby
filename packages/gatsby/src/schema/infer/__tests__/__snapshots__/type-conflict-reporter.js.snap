// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Type conflict reporter prints conflicts and their locations 1`] = `
Array [
  " - type: number
   value: [33m1[39m
   source: Foo Node
",
]
`;

exports[`Type conflict reporter prints conflicts and their locations 2`] = `
Array [
  " - type: boolean
   value: [33mtrue[39m
   source: Foo Node
",
]
`;

exports[`Type conflict reporter prints conflicts and their locations for nested objects 1`] = `
Array [
  " - type: number
   value: [33m1[39m
   source: Foo Node
",
]
`;

exports[`Type conflict reporter prints conflicts and their locations for nested objects 2`] = `
Array [
  " - type: boolean
   value: [33mtrue[39m
   source: Foo Node
",
]
`;
